src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,163 @@
+            Node<T> previousNode = soughtForNode.prev;
+            Node<T> newNode = new Node<>(value, soughtForNode, previousNode);
+            soughtForNode.prev = newNode;
+            if (previousNode == null) {
+                firstNode = newNode;
+            } else {
+                previousNode.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list.size() == 0) {
+            return false;
+        }
+        Iterator iterator = list.iterator();
+        while (iterator.hasNext()) {
+            add((T) iterator.next());
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> soughtForNode = findNodeByIndex(index);
+        return soughtForNode.value;
     }
 
+    // DOES THIS METHOD REPLACE THE NODE OR SHIFT IT?
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToBeShifted = findNodeByIndex(index);
+        T previousValue = nodeToBeShifted.value;
+        nodeToBeShifted.value = value;
+        return previousValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToBeRemoved = findNodeByIndex(index);
+        removeNodeLinks(nodeToBeRemoved);
+        return nodeToBeRemoved.value;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> nodeToBeRemoved = firstNode;
+        for (int i = 0; i < size; i++) {
+            if (nodeToBeRemoved.value == t
+                    || nodeToBeRemoved.value != null && nodeToBeRemoved.value.equals(t)) {
+                removeNodeLinks(nodeToBeRemoved);
+                return true;
+            }
+            nodeToBeRemoved = nodeToBeRemoved.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> soughtForNode;
+        if (index < size >> 1) {
+            soughtForNode = firstNode;
+            for (int i = 0; i < index; i++) {
+                soughtForNode = soughtForNode.next;
+            }
+        } else {
+            soughtForNode = lastNode;
+            for (int i = size - 1; i > index; i--) {
+                soughtForNode = soughtForNode.prev;
+            }
+        }
+        return soughtForNode;
+    }
+
+    private void removeNodeLinks(Node<T> nodeToBeRemoved) {
+        Node<T> previousNode = nodeToBeRemoved.prev;
+        Node<T> nextNode = nodeToBeRemoved.next;
+        if (previousNode == null) {
+            firstNode = nextNode;
+        } else {
+            previousNode.next = nextNode;
```

Comment:
in this task it is important to understand the concept of links
so let`s remove the links from both sides